### ISSUE TITLE
_get_bin takes hours with queue size 1M.

### ISSUE THREAD

**Issue by yrik** [istresearch/scrapy-cluster#53]
I'm scraping etsy.com and queue size  become more than 1M. When I query for info/statistics it stuck on `_get_bin` function in `scrapy-cluster/redis-monitor/plugins/info_monitor.py` file. Also i takes 500MB memory for redis-monitor in that moment.

1) What is the best way to keep queue size small?
2) Perhaps `_get_bin` should be rewritten in more efficient way to calc statistics in the database.

**Comment by madisonb** [comment]
1) Best way to keep queue size small? I would recommend reading the [Production Setup](http://scrapy-cluster.readthedocs.org/en/latest/topics/advanced/productionsetup.html) or how to [Control](http://scrapy-cluster.readthedocs.org/en/latest/topics/crawler/controlling.html) your cluster for quicker crawling. In an idea world you set your crawl limit below the website limit that will get you banned, or spread your crawlers across many machines (but still under the limit).
2) In looking at the `_get_bin()` function, it uses a [zscan](http://redis.io/commands/scan) iterator over the queue, which is the recommended way to iterate over large keys within Redis.

If you would like, I will put a couple million crawl requests into the virtual machine queue and see how the redis monitor fairs.

**Comment by yrik** [comment]
Thanks for reply.

1. Checked it out several times... I have only one machine where I run it. Website does not ban me. I used standard "link" crawler. The issue is that pending pages list constantly growing and eats lots of memory  + makes stat calculation per appid very slow (hours). Just don't understand how it possible to control. For example crawler goes to a page that contains 30 links, so those 30 automatically going to pending list. With a site where 35M of pages I assume that there will be a moment when at least 10M links will be in memory? No way to limit it?

2. Think it needs debugging as issue is definitely there. At least in my case it takes hours to process one info request. I improved the time a bit by adding rows like 
```
if 'meta' in my_item:
    my_item = my_item['meta']
my_item = {'appid': my_item.get('appid'), 'crawlid': my_item.get('crawlid'), 'priority': my_item.get('priority')}
```
to keep in memory only params that is used in next loops.
But it still slow.

**Comment by madisonb** [comment]
1. For point one, my reference to the above was to increase the rate your spiders crawl the website, or to add more computers to your crawling cluster. I assume you submitted a crawl request of depth 2 or 3 (or even higher), and yes, that will generate you millions of pages to crawl. By controlling, you can either scale out the number of spiders you run, alter the domain throttle, or alter the overall cluster throttle -> that will make your spiders crawl faster. The only reason new links are added is if you specifically asked for it.

2. I currently have 200,000 requests in my local vm (4G of ram) queue and see almost no degradation in performance for any of the stats dumps or info requests. The only time I have performance issues is when I try to do an `info` request with the `appid` that spans all of the craw requests, but even then it is maybe 5 min delay (so plausible delay of hour). I have unique crawlid's for all my requests and getting the `info` for a particular crawl id takes around 20 seconds. This isnt that bad considering we have to iterate over 200,000 items within redis.

I certainly agree with your optimization in that we dont need to store the whole crawl request in memory, but building the dictionary required for the `info` request is still computationally expensive.

I understand you are running this on one machine, but why? If you are crawling 1 million pages your are generating approximated 291G of data (esitmate of 2kb per page), 35 million pages is going to be in the terabytes level easily. If you are doing that much crawling, you should have a processing infrastructure of many machines just to get your data, and equally as large to actually process it and store it.

**Comment by madisonb** [comment]
I would also note that in the future, you can use the `expires` field on your crawl request to set a dedicated expiration for your crawl (so it does not take too long or too much time). OR you can send a `stop` action request to the cluster to immediately halt your `crawlid`.

In thinking about this more, I think the main issue here is _how large_ your crawl queue has grown, which can be fixed in a number of different ways:
- Add more crawlers to your local machine
- Increase the standard cluster throttle -> the default is 10 requests every minute per IP address, applies to ALL domains
- Increase the domain specific throttle, similar to above but applies to only 1 domain
- Increase your network bandwidth, as in buy more machines or upgrade to a better package

In production we have enough crawler machines and they are configured properly that our domain queue only gets to maybe 100 or 200 requests deep in the backlog. We can burn through 1 or 2 million requests in less than a day, depending on how high a priority the crawl is for our needs. That requires money and servers, which is not always available. Keep that in mind when you are trying to execute huge crawls on your laptop or a single server.

**Comment by yrik** [comment]
Thank you for a detailed response.

Just don't understand one simple thing.
When we increase crawling speed, we just grow queue faster, as every page contains at least several links to other pages. In case of big site most of new links will be unique. Is is correct? Is the only way to crawl big site  to have huge queue in memory?

At the moment I don't store results of scrapping, it's rather test of crawling speed and how can I control server load. So I don't have an issue with the terabyte of data. It processed 0.3M pages in a day and queue became 2M with `QUEUE_HITS = 60`.

Theoretically, on limited resources, it should just take longer, but not go into the crash mode.

**Comment by madisonb** [comment]
For crawl depth of greater than 0, I like to think of the next layer deep as being exponential, like somewhere between ~ 10^x to 20^x, where x is the depth you are crawling. If there are 20 links per page, you are at 3.2million just by depth 3.

The link spider by default crawls all links at 1 layer, and decrements the next layer by a priority of 10. If you remove that decrement, all your links would be on the same priority level and I think Redis would just give you random items from that level, so your queue may not grow as big. Also, if you added 10 instead of subtracted 10, you would do a depth first crawl instead of a breadth first, which may be the best option for you to reduce your crawl queue size.

Regardless, increasing the number of crawlers and IP addresses will help you burn through the queue faster - there is no way to circumvent the queue backlog, so you might as well crawl it as fast as you can. For each IP address you have, you can multiple your crawl rate by a factor of 1. 2 IPs = 2x the crawl rate, 3 IPs= 3x the crawl rate. This is all configurable for how you want it to be.

Lastly, there is a Queue Stat dump by the Redis Monitor, have you looked at that log file? That gives you the backlog of every domain queue for every spider, regardless of appid or crawlid. Issue #40 tracks this, as we dump it to the logs but is not part of the API yet. 

Hope all this helps you figure out what will work best for your needs.

**Comment by yrik** [comment]
>Also, if you added 10 instead of subtracted 10, you would do a depth first crawl instead of a breadth first, which may be the best option for you to reduce your crawl queue size.

Thanks, it's a great hint!

>Lastly, there is a Queue Stat dump by the Redis Monitor, have you looked at that log file?
Thanks, already using it to track stats per crawler.

Thanks again for all the sugestions.

**Comment by madisonb** [comment]
On a remaining note - I still think the time delay is an issue but I am unsure how to solve it without iterating over all requests in redis. Also, we are using the `pickle` library to serialize the requests in Redis, and according to [this](https://gist.github.com/cactus/4073643#file-results-txt-L24) is insanely slow. So we may need to flip over everything that says `pickle` to a more robust serialization library like `cbor`, `ujson`, or `marshal`.

At the very minimum I think that needs to be addressed considering the performance implications.

**Comment by joequery** [comment]
My vote is for a non-binary serialization format just so the data within redis is still easily accessible and readable outside of scrapy-cluster (or any application). The fastest non-binary serialization library listed in the benchmarks is ujson. [Messagepack](http://msgpack.org/index.html) does seem really interesting if we're going to go the binary format route, and redis supports it out of the box.

**Comment by madisonb** [comment]
Flipping our serialization to ujson does indeed make this process faster, but there are still considerations to be made:

1) Your resulting Json Object may be too large to fit into Kafka. - This is due to the way the info monitor data is structured, where the hierarchy goes `{appid: {crawlid: }`, so having 100,000 different crawlids results in 100,000 different objects that look like:

```
u'b085a8e6-27cc-46ff-aab2-9cae16ae6165': {'domains': {'istresearch.com': {'low_priority': 1, 'high_priority': 1, 'total': 1}}
```
Which causes a pretty big spike in memory usage. This can be mitigated pretty easily by writing a different Redis Monitor plugin that constructs the data how you want it to look.

2) There is still a heavy amount of network traffic. - Judging by the fact that a specific crawlid info request is now pretty fast, I am not so concerned about this, but when transferring 1,000,000 records over the wire we still need to consider some latency

3) Memory footprint. I can see the the reasoning behind the issue here with the memory footprint within `_get_bin()`, and I added some variations of suggestion above but it actually causes the script to be slower in my testing. I presume the complier has optimized it the best it can.

----

After writing this all up, and in trying various optimizations, I realized there is one critical bottleneck: https://goo.gl/gi8bXB . For every crawlid, we not only test a call back to redis if it exists, but also have to grab it too. This data already exists within the object, and we don't have to make the redis calls!

Iterating over my sample 230,000 request set takes now around 10 seconds, which is well within the tolerance I think we are looking for. 1 million requests should now be under a minute response time, here is an example log:

```
2016-04-25 00:45:26,222 [redis-monitor] INFO: Received info request
2016-04-25 00:45:35,865 [redis-monitor] INFO: Sent info to kafka
```
